text/0000-build-systems.md

```diff
@@ -0,0 +1,446 @@
+node in the dependency graph into a finer-grained set of nodes for running rustc
+etc.
+
+Finally, there's the actual build *execution*, which is conceptually
+straightforward: we analyze the dependency graph and existing, cached artifacts,
+and then actually perform any un-cached build steps (in parallel when
+possible). Of course, this is the bread-and-butter of many external build
+systems, so we want to make it easy for them to tweak or entirely control this
+part of the process.
+
+The first two steps -- dependency resolution and build configuration -- need to
+operate on an entire dependency graph at once. Build lowering, by contrast, can
+be performed for any crate in isolation.
+
+### Cargo plugins
+
+**A key point is that, in principle, each of these steps is separable from the
+others**. That is, we should be able to rearchitect Cargo so that each of these
+steps is managed by a distinct component, and the components have a stable --
+and public! -- way of communicating with one another. That in turn will enable
+replacing any particular component while keeping the others.
+
+More concretely, this RFC proposed that we pursue *some* kind of plugin system
+in which you can inform Cargo that part of its normal work should instead by
+done by an external tool. This might live in `Cargo.toml`, or in Cargo
+configuration. In any case, each phase of Cargo's work should be swappable.
+
+Note that the four-step division above is a useful conceptual model, but is
+ultimately too coarse-grained; we will likely need to divide up Cargo's
+functionality into numerous small pieces that can be re-used when integrating
+into a larger build system. This finer division is left as a question for
+experimentation.
+
+## Specifics for the homogenous build system case
+
+For homogenous build systems, there's a key question: how is the Rust code
+itself managed, through a crate registry or though some external system? Any
+integration has to handle the first case (to have access to crates.io or a
+mirror thereof), but organizations can choose whether to manage their own crates
+through a custom registry (more on that below) or some other means.
+
+### Using crates managed by a crate registry
+
+Using a crate registry implies using Cargo's dependency resolution, and, in
+particular, `Cargo.toml`. In this case, the external build system should invoke
+Cargo for *at least* the dependency resolution and build configuration steps,
+and likely the build lowering step as well. In such a world, Cargo is
+responsible for *planning* the build (which involves largely Rust-specific
+concerns), but the external build system is responsible for *executing* it.
+
+One important concern is: how do you depend on code from other languages, which
+is being managed by the external build system? That's a narrow version of a more
+general question around *native dependencies*, which will be addressed
+separately in a later section.
```

Comment:
Hmm, I'm not sure the later section really addresses how homogeneous build systems will tend to use native dependencies. That section seems to be restricted to how packages on crates.io specify native dependencies. That's an important part of the story, of course. But ideally, authors of Rust crates within Facebook's monorepo would not get Cargo involved in their native dependencies at all. They would use standard Buck rules to specify their native dependencies, and Buck would build those deps and provide libraries that Rust could link against.

Reply:
Hmm... That's very similar to what I've been doing for OS projects to link the kernel with external assembly code, except the other way around. I have found it easier to compile my rust code into a `.a`, compile my other stuff into `.o`s and pass them all to `ld`.

The annoying part of this approach is that if you want to use any other rust libraries (e.g. `liballoc` or `libcollections`) you also need to obtain them and pass them to either rustc or ld... It gets to be a bit messy if your not careful.

Reply:
I've updated this section to try to make things more clear. The setup is really very simple: Cargo/rustc don't need to know much at all about these native deps, other than their location. The build system manages the rest.